### ISSUE TITLE
Feed returning no items

### ISSUE THREAD

**Issue by unlikenesses** [simplepie/simplepie#474]
This URL is returning no items in SimplePie 1.3.1:

[http://www.burberryplc.com/rss/rssFeed.aspx?type=press](http://www.burberryplc.com/rss/rssFeed.aspx?type=press)

Could it be the query string? [This issue](https://github.com/simplepie/simplepie/issues/119), from 2011, says it was resolved in 1.2.2. But [this issue](https://github.com/simplepie/simplepie/issues/317) says the problem is still occurring in 1.3.1.

**Comment by mblaney** [comment]
Hi @unlikenesses your feed has a Byte Order Mark at the start of the file: \xef\xbb\xbf and it looks like SimplePie can't handle that. (See Content/Type/Sniffer.php: feed_or_html) would probably be a good thing to support.

**Comment by unlikenesses** [comment]
Thanks @mblaney. Is there a quick and dirty hack I can use to allow SimplePie to ignore the BOM?

**Comment by mblaney** [comment]
Sure just set your own content type header to something other than text/html so that SimplePie doesn't have to guess.

**Comment by unlikenesses** [comment]
Thanks, but I'm not in control of what feed URLs are passed to my SimplePie installation - it's dynamic and could be anything, not just the example feed above. I was wondering if there's a hack to Sniffer.php that would accept the BOM?

**Comment by JanPetterMG** [comment]
If you could download the feed yourself, and then pass the raw xml to simplepie, this fix _might_ work (untested):
```php
$str = str_replace("\xEF\xBB\xBF",'',$str); 
```

**A much better solution:**
I would highly recommend switching to [GuzzleHttp](https://github.com/guzzle/guzzle). Let Guzzle download everything for you, then Simplepie can do what it was made to do, parsing feeds.
It's easy to fix, and it doesn't require many changes either.

Working example, no more BOM issues:
```php
$guzzle = new \GuzzleHttp\Client();
$response = $guzzle->request('GET', 'http://www.burberryplc.com/rss/rssFeed.aspx?type=press');
$xml = $response->getBody()->getContents();

$simplepie = new \SimplePie();
$simplepie->set_raw_data($xml);
$simplepie->init();
```
The content is automatically decoded (by default). More options like timeout, useragent, redirects, when to fail, ssl/tls certficate verification etc, is configurable as an input array in the constructor.

**Comment by mblaney** [comment]
hey @JanPetterMG I would love to see SimplePie split into sections like that, and then compare the way it pulls content compared to Guzzle. It would be great to understand where caching, 304's, etc would fit in to that model.

@unlikenesses yes you can hack Sniffer.php to do that, just change this line at the start of feed_or_html: `$pos = strspn($this->file->body, "\x09\x0A\x0D\x20\xEF\xBB\xBF");`

**Comment by JanPetterMG** [comment]
@mblaney Yeah, using my above code would eliminate any 304's, resulting in slightly increased traffic, witch isn't optimal, but hey, not all hosts do support/send 304's.
Simplepie's caching part is optional right? So when this is not used, there wouldn't be any differences in the amount of load, only benefits.

This might not be the right place to discuss this, but if we could migrate to Guzzle, I would gladly contribute. It will simplify Simplepie a lot, as well as fixing some existing issues. For example PHP 7 builds still fails, mostly due to bugs related to URL parsing, and it will include an BOM fix out of the box. Additionally, large bunches of code isn't needed anymore, witch means easier to maintain.

An Guzzle integration in Simplepie is possible, could be released as v1.5 without breaking any backwards support. We could also implement it side-by-side, should in theory be easier, but it's probably not...
For those who isn't using Composer already (wordpress users?), we could include an copy of GuzzleHttp v6 in the zip file.
What do you think?

**Comment by mblaney** [comment]
@JanPetterMG I think that sounds great, but I also don't think there's enough interest in SimplePie anymore (or maybe just at the moment), to warrant re-factoring. If you're keen to do it though go for it! I'm happy to look at your changes, and if others are happy maybe create an official (experimental) v2 branch here? It's probably worth keeping versions 1.x for smaller changes.

**Comment by unlikenesses** [comment]
Thanks for the advice @JanPetterMG. I like the sound of using Guzzle, but at the moment I use SimplePie's `set_feed_url` to handle an array of feeds. I guess your method would not allow any merging `set_feed_url` performs.

**Comment by JanPetterMG** [comment]
@unlikenesses That's right.

I'm just wondering why BOM support never was implemented. Anyone knows?

How about checking the first character, and match it against [this](https://en.wikipedia.org/wiki/Byte_order_mark#Representations_of_byte_order_marks_by_encoding) table?
I guess Simplepie (currently) ignores any url content, if the it doesn't start with the character `<`. Am I onto something now?

**Comment by mblaney** [comment]
fixed in #486
